### PR TITLE
group sources upgrades

### PR DESCRIPTION
**Pull request by guynir42**
Includes filterable classifications, TNS name (needs to be finished), date saved, separating decimal/sexagesimal coords and only displaying one of them, fixing the CVS output. 
group_sources_upgrades

### CODE REVIEW

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:264]: Use `source.saved_at`

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:233]: Only show the most recent classification in this table

**Comment by dmitryduev** [comment]
@guynir42 this looks great! May I request one thing? I would make the table width be 100% of the available screen space regardless of the amount of content, like on the Candidates page. (If there's more stuff, horizontal scrolling if perfectly fine, I think) @dannygoldstein @kmshin1397 you guys agree?

File: static/js/components/GroupSources.jsx
Comment (stefanv) [inline static/js/components/GroupSources.jsx:241]:
```suggestion
```

**Review by stefanv — COMMENTED** [review]
Needs tests, otherwise looks great!

File: static/js/components/GroupSources.jsx
Comment (kmshin1397) [inline static/js/components/GroupSources.jsx:205]: Might be a good idea to specify .toFixed() values for these numerical columns (RA, Dec, redshift) - we don't have to round too aggressively (could be like .toFixed(6) maybe?) but doesn't hurt to just have that in there in case the values are very long I think.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:233]: The ShowClassification component shows only the latest classification...

Comment (kmshin1397) [reply static/js/components/GroupSources.jsx:233]: Agreed - the ShowClassification component will display the most recent classification for each taxonomy, not just overall. 

I think the best way to do this would be to refactor the ShowClassification component slightly to have an prop for toggling showing just the most recent classification vs for each taxonomy. It seems like something we could want to reuse later. Honestly, maybe just use the `shortened` flag you already put in and say, if shortened, only display the most recent classification.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:264]: That does not exist. As far as I can tell, for sources "created_at" represents when the object was saved.

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:263]: I was under the impression that this is going to come in very soon. 

@dmitryduev do you know when we can expect to have TNS names? or at least where I can find them in them when they appear?

**Review by kmshin1397 — COMMENTED** [review]
Looks good overall! A couple of minor comments, and probably could use a test or two as Stefan mentioned.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:233]: I just wanted to leave the behavior of ShowClassifications as it was for other people using it. I agree that we can make this component more versatile. I will have to make sure it doesn't appear in too many other places where you'd suddenly get all classifications instead of one.

File: static/js/components/GroupSources.jsx
Comment (acrellin) [reply static/js/components/GroupSources.jsx:264]: https://github.com/skyportal/skyportal/blob/9c986c12407b9cf418088d7a2b29965b140fc580/skyportal/models.py#L787

**Comment by guynir42** [comment]
I can't get the table to fill the entire screen, some help is needed. 

Also, it looks like the handler for sources actually returns Obj and not Source (at least, when getting multiple sources). I'm not sure how to fix this (see source.py around line 294)

**Comment by kmshin1397** [comment]
> I can't get the table to fill the entire screen, some help is needed.
> 
> Also, it looks like the handler for sources actually returns Obj and not Source (at least, when getting multiple sources). I'm not sure how to fix this (see source.py around line 294)

Hi Guy, I started playing around with this and I think I fixed the width issue - I can just make a PR to your branch give me a few minutes.

**Comment by kmshin1397** [comment]
@guynir42 Opened a PR to fix the width and the saved_at issues [here](https://github.com/guynir42/skyportal/pull/1)

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:384]: I'm not sure what the utility of a list of datetimes is here decoupled from the actual groups they were saved to. I think the saved_at value should be added as a new key to each dict in source_list[-1]["groups"] after that's added on L401 below (will need to call to_dict() on each group beforehand)

**Review by acrellin — CHANGES_REQUESTED** [review]

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:103]:
```suggestion
```

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [reply static/js/components/GroupSources.jsx:264]: On the group sources page people will only be interested in the classification in the context of their group. Please refactor this to show only the latest classification from the group that is being shown

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:379]: add a line break beetween each group chip

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:383]:
```suggestion
      name: "Date Saved",
```

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:300]: Dont render fractions of a second here - dont need that much precision

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:256]: Dont use fixed point here - implies a redshfit precision which may not be achieved in reality. I.e., we dont want to render a redshift that is 0.06 as 0.060000

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:298]: Shouldn't there be only one saved_at value for each GroupSource row?

Comment (acrellin) [reply skyportal/handlers/api/source.py:298]: Probably should be a scalar value for each group rather than a list

Comment (acrellin) [reply skyportal/handlers/api/source.py:298]: Also, query.all() already returns a list -- here you're wrapping it in another list

Comment (acrellin) [reply skyportal/handlers/api/source.py:298]: I think you can scrap the list comprehension and use `.first()`

Comment (guynir42) [reply skyportal/handlers/api/source.py:298]: I think we talked about having the ability to save the same source to the same group more than once. 

In any case, it doesn't hurt to have it as an array if we ever do decide that is an option.

Comment (acrellin) [reply skyportal/handlers/api/source.py:298]: That's not even possible with the current data model. There can only be one GroupSource row for a given obj_id, group_id pair, and saved_at is necessarily a single scalar value. Please update this to reflect the current data model. If we ever refactor that in the future, we can update this accordingly. Thanks

**Comment by dannygoldstein** [comment]
merge conflicts

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py]: OK

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:443]: This is great! Just need to remove the commented out code above

**Review by dannygoldstein — CHANGES_REQUESTED** [review]
@guynir42 API test is failing on Travis and GA - this failure is real

**Review by dannygoldstein — APPROVED** [review]

File: skyportal/tests/frontend/test_group_sources.py
Comment (acrellin) [inline skyportal/tests/frontend/test_group_sources.py:135]:
```suggestion
```

File: skyportal/tests/frontend/test_group_sources.py
Comment (acrellin) [inline skyportal/tests/frontend/test_group_sources.py:21]:
```suggestion
```

File: skyportal/handlers/api/source.py
Comment (acrellin) [reply skyportal/handlers/api/source.py:292]:
```suggestion
```
Instead of an empty string placeholder, we should simply be able to handle the case where this is null or undefined on the front-end gracefully

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:324]: A simplification:
```suggestion
            for i in range(len(source_info["groups"])):
                saved_at = (
                    DBSession()
                    .query(Source.saved_at)
                    .filter(Source.obj_id == source_info["id"], Source.group_id == g.id)
                    .first()
                    .saved_at
                )
                g_new = g.to_dict()
                g_new['saved_at'] = saved_at
                source_info["groups"][i] = g_new
```

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:422]:
```suggestion
```

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:441]: Same simplification request here as above

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:238]:
```suggestion
  const renderDecSex = (dataIndex) => {
```

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:227]:
```suggestion
  const renderRASex = (dataIndex) => {
```

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:413]: Can you please add a note explaining why an empty string is required here for the finder chart?

File: static/js/ducks/sources.js
Comment (acrellin) [inline static/js/ducks/sources.js:102]: I don't see this WS message type ever being pushed on the backend -- why is this needed?

**Review by acrellin — CHANGES_REQUESTED** [review]
Looks great overall @guynir42 , just a few requests/comments/questions below

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:324]: enumerate, pretty cool :)

File: static/js/ducks/sources.js
Comment (guynir42) [reply static/js/ducks/sources.js:102]: I used it when I wanted the group sources page to update automatically. 

I guess we have decided to make it non-reactive, so I can remove this handler...

**Comment by acrellin** [comment]
Hey @guynir42, if this is ready for re-review, please indicate that with the re-request review button, otherwise I/we don't know to take another look. I see that some recently merged changes created conflicts - let me know if you'd like a hand resolving those. Thanks!

File: skyportal/tests/api/test_candidates.py
Comment (acrellin) [inline skyportal/tests/api/test_candidates.py:16]:
```suggestion
    status, _ = api(
```
Just a two-element tuple returned by `api`

**Review by acrellin — APPROVED** [review]
Looks great!

File: skyportal/tests/api/test_candidates.py
Comment (guynir42) [reply skyportal/tests/api/test_candidates.py:16]: What happens if it returns a scalar? or is that never supposed to happen?

Comment (acrellin) [reply skyportal/tests/api/test_candidates.py:16]: I just mean the naming convention for unused variables. It's fine as-is, but I think `*_` is typically used to indicate multiple unused values, whereas `_` is used for single values (in this case a dict)

Comment (guynir42) [reply skyportal/tests/api/test_candidates.py:16]: I think *_ is preferred here because if there are no additional outputs beyond the first then it is ok, and if there are multiple additional outputs it is also ok. 
If you put status, _ = ... it expects exactly two outputs.

Comment (acrellin) [reply skyportal/tests/api/test_candidates.py:16]: `api` always returns at most two outputs

Comment (guynir42) [reply skyportal/tests/api/test_candidates.py:16]: Yes, but if it returns only one, and you specify status, _ = ... then it doesn't work, I think...

**Comment by acrellin** [comment]
Will merge when green

File: skyportal/tests/api/test_candidates.py
Comment (stefanv) [reply skyportal/tests/api/test_candidates.py]: I propose that we always return two outputs, making it consistent for `HEAD` requests.

**Comment by stefanv** [comment]
@acrellin Before (or after) merging, please check that baselayer is not downgraded.  Some PRs pin old versions of baselayer.

**Comment by acrellin** [comment]
> @acrellin Before (or after) merging, please check that baselayer is not downgraded. Some PRs pin old versions of baselayer.

I just pushed up a commit pinning the latest baselayer

File: skyportal/tests/api/test_candidates.py
Comment (acrellin) [reply skyportal/tests/api/test_candidates.py]: Agreed

**Comment by acrellin** [comment]
Failures appear unrelated: 
```
FAILED skyportal/tests/frontend/test_scanning_page.py::test_submit_annotations_filtering
FAILED skyportal/tests/frontend/test_sources.py::test_upload_download_comment_attachment
```

**Pull request merged**
